File: data/Queue.ts
export default class Queue<T> {
  private items: T[] = [];
  private max?: number;

  constructor(max?: number) {
    this.max = max;
  }

  public push(item: T) {
    if (this?.max !== undefined && this.items.length >= this.max) {
      return false;
    }

    this.items.unshift(item);
    return true;
  }

  public pop() {
    return this.items.pop() ?? null;
  }

  public peek() {
    const slice = this.items.slice(-1);

    if (slice.length === 0) {
      return null;
    }
    return slice[0];
  }

  public size() {
    return this.items.length;
  }

  [Symbol.iterator]() {
    let index = this.items.length - 1;

    return {
      next: () => {
        if (index > -1) {
          const res = { value: this.items[index], done: false };
          index = index - 1;
          return res;
        } else {
          return { done: true };
        }
      },
    };
  }
}

File: data/queue.go
package data

type Queue struct {
	items []interface{}
	max   int
}

func NewQueue(max int) *Queue {
	return &Queue{items: make([]interface{}, 0, max), max: max}
}

func (q *Queue) Push(item interface{}) bool {
	if q.max != 0 && len(q.items) >= q.max {
		return false
	}
	q.items = append([]interface{}{item}, q.items...)
	return true
}

func (q *Queue) Pop() interface{} {
	if len(q.items) == 0 {
		return nil
	}

	item := q.items[len(q.items)-1]
	q.items = q.items[:len(q.items)-1]
	return item
}

func (q *Queue) Peek() interface{} {
	if len(q.items) == 0 {
		return nil
	}
	return q.items[len(q.items)-1]
}

func (q *Queue) Size() int {
	return len(q.items)
}
